Blocks E2E: Refactor configs and workflow

WunderBart opened the pull request:
Follow-up to https://github.com/woocommerce/woocommerce/pull/46125
Closes: #46331
Related: pdV5qK-nT-p2#possible-solutions

## Proposed Changes

This PR enhances the efficiency of running our Blocks E2E tests both locally and in our CI environment.

First, we have eliminated all `side effects` references as a continuation of #46125, ensuring they are permanently resolved. 🎉 

Regarding CI improvements, we've restructured the Playwright configuration (#46331) to enable all tests to run in a single job rather than a 3D theme-based matrix. This adjustment reduces resource usage by half (10 instead of 20 shards) and accelerates test completion due to better-balanced shards (cc: @tammullen). Additionally, we introduced a `Merge Artifacts` step, consolidating all debug artifacts from separate shards into one zip file for convenience.

For local testing improvements, we've reduced global setup rerun times by reusing existing database snapshots created within the initial run (introduced in #46125). We've also simplified the test execution process, making it unnecessary to identify the specific project (`block-theme`, `classic-theme`, or `block-theme-with-templates`) the test belongs to. Running a single suite is now as straightforward as using `pnpm test:e2e path-to/any-suite.spec.ts`. However, executing all theme-based groups remains an option through `pnpm test:e2e:block-theme/class-theme/block-theme-with-templates` when required.

## Testing Instructions

Ensure running tests via `pnpm test:e2e`, `pnpm test:e2e:block-theme`, `pnpm test:e2e:classic-theme`, `pnpm test:e2e:block-theme-with-templates` works as expected. Ensure all CI tests pass.

⚠️ Some tests might still be flaky. However, that flakiness is out of the scope of this PR, as it is no longer caused by the side effects. Those flaky tests need to be investigated individually.

## Next Steps

- Update [E2E docs](https://github.com/woocommerce/woocommerce/blob/8ee74ede0ebc1957089af08207ae2a1a8250610e/plugins/woocommerce-blocks/docs/contributors/e2e-guidelines.md).
- Explore uploading DB snapshot to the repo to reduce setup time.

WunderBart committed:
Merge all e2e configs

WunderBart committed:
Refactor tests workflow to address config changes

WunderBart committed:
Move db import to page fixture setup
This way we're always sure each test will start from the same initial state

WunderBart committed:
Improve wording

WunderBart committed:
Revert "Move db import to page fixture setup"

This reverts commit 58ebd13ae5e6ab53d50c83c110042a2c514e23cb.

WunderBart committed:
Tweak db import messages

WunderBart committed:
Fix artifacts path

WunderBart committed:
Fix artifacts path again + bump action version

WunderBart committed:
Tweak job title

WunderBart committed:
Tweak some more titling

WunderBart committed:
Fix artifacts name

WunderBart committed:
Use beforeEach hook for activating themes

WunderBart committed:
Some more workflow tweaks

WunderBart committed:
Remove one more hook

WunderBart committed:
Fix failing guest shopper test

WunderBart committed:
Add shard prefix to artifacts upload

WunderBart committed:
Use actions/checkout@v4

WunderBart committed:
Add artifacts merge step

WunderBart committed:
Remove comment

WunderBart committed:
Add changelog entry

WunderBart committed:
Try generating html report

WunderBart committed:
Revert "Try generating html report"

This reverts commit c7f5e347a908d24aa87e194665160ddbb9748389.

WunderBart committed:
Streamline db error

WunderBart committed:
Make exec reject on error

WunderBart committed:
Try 10 shards

WunderBart committed:
Execute templates tests in parallel like before

WunderBart committed:
Revert "Execute templates tests in parallel like before"

This reverts commit e215832eac61184b542db5ef33e001a48ced2b0b.

WunderBart committed:
Remove side_effects suffix from test files

WunderBart committed:
Remove accidental pause

WunderBart committed:
Fix linter errors

WunderBart committed:
Restore the themed runners

WunderBart committed:
Reuse db snapshot to reduce global setup time

Aljullu commented on plugins/woocommerce-blocks/tests/e2e/playwright-utils/test.ts:
```diff
@@ -161,12 +161,9 @@ const test = base.extend<
+		await cli(
 			`npm run wp-env run tests-cli wp db import ${ DB_EXPORT_FILE }`
 		);
-		if ( ! cliOutput.stdout.includes( 'Success: Imported ' ) ) {
-			throw new Error( `Failed to import ${ DB_EXPORT_FILE }` );
-		}
 	},
```
Out of curiosity: why this change?

WunderBart replied:
Nice catch! That check was needed because the `cli` utility always resolved the promise, regardless of whether the executed script threw an error. I made it reject on an error in https://github.com/woocommerce/woocommerce/pull/46409/commits/3a60fac18769ef218a0157ebf451f363ff6db73c, so now we don't need these explicit output checks as they're handled inside the utility.

Aljullu reviewed (approved):
Once again, thanks for all the improvements in our e2e suite, @WunderBart! :clap: 

I reviewed the code and tried some test commands, and everything looks good on my end. I left one question, but no need to change anything, I'm just curious. :slightly_smiling_face: 

I'm approving the PR, but given that it's a big one and I'm not an expert with e2e tests, I would appreciate if somebody else could take a second look just in case.

gigitux commented on .github/workflows/blocks-playwright.yml:
```diff
@@ -26,47 +25,44 @@ jobs:
+                  name: failures-artifacts-shard-${{ matrix.shardIndex }}
+                  path: plugins/woocommerce-blocks/tests/e2e/artifacts/test-results
+                  if-no-files-found: ignore
+    merge-artifacts:
+        name: Merge Artifacts
+        if: ${{ !cancelled() }}
+        needs: [blocks-playwright-tests]
```
Great idea! ❤️

gigitux reviewed (approved):
Great work on the changes! LGTM!
One area for discussion:  Since the suffix approach is no longer required for environment setup, could we simplify the suffixes to "classic theme" and "block theme"? This would remove the redundant "block_theme_with_templates" suffix, but at the same time, make it easier to identify the theme during E2E tests: what do you think?

WunderBart commented:
> Since the suffix approach is no longer required for environment setup, could we simplify the suffixes to "classic theme" and "block theme"? This would remove the redundant "block_theme_with_templates" suffix, but at the same time, make it easier to identify the theme during E2E tests: what do you think?

I need to figure out if I follow. IIUC, the suffixes are used to separate the themes we test against, which means we currently test against three different themes. What should we do with the suites now tagged with `block_theme_with_templates` if we remove the suffix?

gigitux commented:
> > Since the suffix approach is no longer required for environment setup, could we simplify the suffixes to "classic theme" and "block theme"? This would remove the redundant "block_theme_with_templates" suffix, but at the same time, make it easier to identify the theme during E2E tests: what do you think?
> 
> I need to figure out if I follow. IIUC, the suffixes are used to separate the themes we test against, which means we currently test against three different themes. What should we do with the suites now tagged with `block_theme_with_templates` if we remove the suffix?

True, but other tests using a child theme starting from a classic theme and/or a block theme: https://github.com/woocommerce/woocommerce/blob/f339aee5b58e37e3d79229a176289a49fd925a63/plugins/woocommerce-blocks/tests/e2e/utils/constants.ts/#L6-L17


I would target them with the `.block_theme` because, in the end, `theme-with-woo-templates` is a block theme.

WunderBart commented:
> > > Since the suffix approach is no longer required for environment setup, could we simplify the suffixes to "classic theme" and "block theme"? This would remove the redundant "block_theme_with_templates" suffix, but at the same time, make it easier to identify the theme during E2E tests: what do you think?
> > 
> > 
> > I need to figure out if I follow. IIUC, the suffixes are used to separate the themes we test against, which means we currently test against three different themes. What should we do with the suites now tagged with `block_theme_with_templates` if we remove the suffix?
> 
> True, but other tests using a child theme starting from a classic theme and/or a block theme: https://github.com/woocommerce/woocommerce/blob/f339aee5b58e37e3d79229a176289a49fd925a63/plugins/woocommerce-blocks/tests/e2e/utils/constants.ts/#L6-L17
> 
> I would target them with the `.block_theme` because, in the end, `theme-with-woo-templates` is a block theme.

Thanks for clarifying. Looking even further, do we even need the block/classic theme distinction? Do you think it's a useful thing to have? Previously, it was necessary to make the theme-based split since themes were being activated in the global setup phase. On this branch, we activate within the `beforeEach` hook, so the separation is unnecessary. I'm curious if anyone would actually be using the `pnpm test:e2e:classic-theme/block-theme` command since we usually don't run more than one suite locally, which can be done with `pnpm test:e2e path/to/suite.spec.ts`.

WunderBart committed:
Remove more unnecessary setup and teardown steps

gigitux commented:
> looking even further, do we even need the block/classic theme distinction? Do you think it's a useful thing to have?

Yeah, good point. I think that the classic/block suffixes are useful to have some separation. Otherwise, there is the risk that we will put everything in one file, and E2E tests will be huge.

WunderBart commented:
> > looking even further, do we even need the block/classic theme distinction? Do you think it's a useful thing to have?
> 
> Yeah, good point. I think that the classic/block suffixes are useful for some separation. Otherwise, there is the risk that we will put everything in one file, and the E2E tests will be huge.

I also lean towards keeping them in separate files – it keeps the structure flat. I'll merge the `block-theme-with-templates` with `block-theme`, then!

nielslange commented:
@WunderBart Firstly, thank you so much for working on this! This is another great improvement to the e2e tests. Regarding using different themes, what you and @gigitux discussed above, I recently created https://github.com/woocommerce/woocommerce/issues/46746. Not directly related to the issue raised by @gigitux, but a potential speed improvement when setting up the test env.

WunderBart committed:
Merge remote-tracking branch 'origin' into refactor/e2e-playwright-configs

WunderBart committed:
Refactor out the isMacOS check

WunderBart committed:
Try generating HTML report take 2

WunderBart committed:
Revert "Try generating HTML report take 2"

This reverts commit 0492dae25fcbb0db2ca03f96f6e373c70c1d4a85.

WunderBart committed:
Fix artifact merge when artifacts unavail. after rerun

WunderBart committed:
tmp: print artifact id

WunderBart committed:
Revert "tmp: print artifact id"

This reverts commit 53a94a66a0b618884a69b5dac312aa9c9da19409.

WunderBart committed:
Merge remote-tracking branch 'origin' into refactor/e2e-playwright-configs

WunderBart committed:
Fix mini cart test
Looks like there was an update to the editor and the blocks commercial is no longer visible

WunderBart committed:
Handle perfect run
After a flawless run (0 failured & flaky tests) no artifacts will be produced, so don't fail the merge artifacts step.

WunderBart committed:
Merge remote-tracking branch 'origin' into refactor/e2e-playwright-configs

WunderBart committed:
Merge remote-tracking branch 'origin' into refactor/e2e-playwright-configs
